### ISSUE TITLE
No such file or directory: './saved_model/0816_adddice/wingsnet_37.pth

### ISSUE THREAD

**Issue by GuobinZhangTJU** [Puzzled-Hui/ATM-22-Related-Work#4]
您好，

非常冒昧打扰您，按照您在Github上的描述，timi团队代码中，Use file train2.py and train.py to train and test the model.

在运行train2.py时发现，当提取训练数据时，需要运行下面这行语句，

bp_weight = np.load(os.path.join(self.root, 'LIBBP', 'distance_bp', name + '.npy'))

可是其中的【'LIBBP', 'distance_bp', name + '.npy'】，并没有找到相关获取程序。
我往上翻了一下，如果我要得到'distance_bp',路径下面的文件就需要运行save_BP_weight_tw():，
但是save_BP_weight_tw():中又需要运行 pred = nibabel.load(os.path.join('./data/LIBBP/preds', name+'.nii.gz'))
然后要得到这个语句里面的/preds就需要运行save_gradients_tw(layer=0)
但是运行save_gradients_tw(layer=0)就需要运行weights_dict = torch.load(os.path.join('./saved_model', 'wingsnet_4.pth'))
可是当前我们是训练过程，是没法得到这个save_model的，不是训练结束后才能得到model么？
您看可不可以帮忙解答一下~非常感谢您~

**Comment by Puzzled-Hui** [comment]
Hi, Thank you for your question.

In Timi team, they adopted a three stage method. 

The loaded weight is the result of the first stage. Please refer to the [challenge paper:](https://www.sciencedirect.com/science/article/pii/S1361841523002177)  for more details, especially the method section.

Best,

**Comment by GuobinZhangTJU** [comment]
Thanks for your reply. I have read the paper four times, but found no training steps, can you help me with the running steps?
I have tried the following steps: LIB weight.py——save_gradients.py——BP enhancement.py——tree_parse.py——train2.py,
the above step i think is reasonable, however i really not find the "weights_dict = torch.load(os.path.join('./saved_model', 'wingsnet_4.pth'))" in save_gradients.py
waiting for your reply ~
thanks~

**Comment by KKoml** [comment]
您好，我也是没有save-model，请问您解决了吗？

**Comment by Puzzled-Hui** [comment]
Hi,

The loaded weight is the result of the first stage, which can be trained locally, and the wingsnet_4.pth is the early epoch results.

Best,

**Comment by TangHuiling** [comment]
您好，您上面提到wingsnet_4.pth是第一阶段训练的早期结果，但是在第一阶段的训练中，train.py文件中有一行    weights_dict = torch.load(os.path.join('./saved_model', 'wingsnet_79.pth'))，
wingsnet_79.pth这个预训练权重文件并没有提供。请问可以提供一下吗？

**Comment by Puzzled-Hui** [comment]
@GuobinZhangTJU @KKoml @TangHuiling

Hi, we will deal with it and update the pretrained weights soon, thank you for your patience.

Best,

**Comment by Puzzled-Hui** [comment]
Dear all,

We have uploaded the pretrained weight: [here](https://drive.google.com/file/d/1hxV8nnB8lkCGsOhjJNxM02LqQuL6WP9D/view?usp=drive_link). 

Alternatively, you could pretrain the weight as the first stage use only Dice loss by yourselves. The early epochs' weight can be used for LIB weight, and the late epochs' weight can be used as the initial weight for the second stage.

Best,

**Comment by GuobinZhangTJU** [comment]
> Dear all,
> 
> We have uploaded the pretrained weight: [here](https://drive.google.com/file/d/1hxV8nnB8lkCGsOhjJNxM02LqQuL6WP9D/view?usp=drive_link).
> 
> Alternatively, you could pretrain the weight as the first stage use only Dice loss by yourselves. The early epochs' weight can be used for LIB weight, and the late epochs' weight can be used as the initial weight for the second stage.
> 
> Best,

Thanks for your reply~
best~

**Comment by elfpattern-code** [comment]
您好，我根据这个步骤LIB weight.py——save_gradients.py——BP enhancement.py——tree_parse.py——train2.py，进行训练，python train2.py，发现模型不收敛，dice_loss一直是Nan，请问这种怎么进行排查？

**Comment by TangHuiling** [comment]
你好~我之前也遇到过这种情况，首先看计算dice_loss的公式，是否存在类似分母为0的情况，其次重点看数据是否为空（如生成的权重文件、预测图像等）希望可以帮到你






唐慧玲
***@***.***






---Original---
From: ***@***.***&gt;
Date: Tue, Nov 19, 2024 14:48 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [EndoluminalSurgicalVision-IMR/ATM-22-Related-Work] No such fileor directory: './saved_model/0816_adddice/wingsnet_37.pth (Issue #4)




 
您好，我根据这个步骤LIB weight.py——save_gradients.py——BP enhancement.py——tree_parse.py——train2.py，进行训练，python train2.py，发现模型不收敛，dice_loss一直是Nan，请问这种怎么进行排查？

**Comment by elfpattern-code** [comment]
刚看了下，都不会空，就第一个epoch数据还正常，第二个之后分子和分母的值就变成了nan，这是为什么？@TangHuiling

**Comment by elfpattern-code** [comment]
![image](https://github.com/user-attachments/assets/5cdee42e-8e13-4efc-8d4b-57dbe3e79f33)
基本上训练都是这种情况，第二次迭代的时候就nan了，咱们这个方法会生成很多中间结果，比如preds，grads，bp_enhancement，weight，怎么去确定这些生成的都是正确的？@GuobinZhangTJU @KKoml @TangHuiling

**Comment by TangHuiling** [comment]
少用几个案例训练，然后把preds,weight这些打印出来一一排查，我当时好像有一个结果全是0和1。












唐慧玲
***@***.***






---Original---
From: ***@***.***&gt;
Date: Tue, Nov 19, 2024 17:01 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [EndoluminalSurgicalVision-IMR/ATM-22-Related-Work] No such fileor directory: './saved_model/0816_adddice/wingsnet_37.pth (Issue #4)




 
image.png (view on web)
 ***@***.*** @KKoml @TangHuiling

**Comment by elfpattern-code** [comment]
![微信图片_20241120110609](https://github.com/user-attachments/assets/f78afccb-8f21-4537-bbff-08ac3754df78)
您好，由于是新人小白，不是很懂，还得麻烦您给看一下。我用4个数据训练了一下，把weight，label，preds这些的sum都打印出来了，有时候label是0，有时候weight是nan，为什么会出现这种情况？@TangHuiling

**Comment by Gweiqi** [comment]
> 您好，
> 
> 非常冒昧打扰您，按照您在Github上的描述，timi团队代码中，Use file train2.py and train.py to train and test the model.
> 
> 在运行train2.py时发现，当提取训练数据时，需要运行下面这行语句，
> 
> bp_weight = np.load(os.path.join(self.root, 'LIBBP', 'distance_bp', name + '.npy'))
> 
> 可是其中的【'LIBBP', 'distance_bp', name + '.npy'】，并没有找到相关获取程序。 我往上翻了一下，如果我要得到'distance_bp',路径下面的文件就需要运行save_BP_weight_tw():， 但是save_BP_weight_tw():中又需要运行 pred = nibabel.load(os.path.join('./data/LIBBP/preds', name+'.nii.gz')) 然后要得到这个语句里面的/preds就需要运行save_gradients_tw(layer=0) 但是运行save_gradients_tw(layer=0)就需要运行weights_dict = torch.load(os.path.join('./saved_model', 'wingsnet_4.pth')) 可是当前我们是训练过程，是没法得到这个save_model的，不是训练结束后才能得到model么？ 您看可不可以帮忙解答一下~非常感谢您~
您好,
冒昧打扰了.我在第一阶段的训练中，train.py 文件有一行 File “train.py”， line 244， in test  weights_dict = torch.load（os.path.join（'./saved_model'， folder_name， 'wingsnet_18.pth'）），报错] 没有这样的文件或目录： './saved_model/0816_adddice/wingsnet_18.pth'，请问这个wingsnet_18.pth是如何获得的？您是否遇到过这样的问题？期盼您的回复，感激不尽

**Comment by Gweiqi** [comment]
> 您好，
> 
> 非常冒昧打扰您，按照您在Github上的描述，timi团队代码中，使用文件 train2.py 和 train.py 来训练和测试模型。
> 
> 在运行 train2.py 时发现，当提取训练数据时，需要运行下面这行语句，
> 
> bp_weight = np.load（os.path.join（self.root， 'LIBBP'， 'distance_bp'， name + '.npy'））
> 
> 可是其中的【'LIBBP'， 'distance_bp'， name + '.npy'】，并没有找到相关获取程序。 我往上翻了一下，如果我要得到'distance_bp'，路径下面的文件就需要运行save_BP_weight_tw（）：，但是save_BP_weight_tw（）：中又需要运行 pred = nibabel.load（os.path.join（'./data/LIBBP/preds'， name+'.nii.gz'）） 然后要得到这个语句里面的/preds就需要运行save_gradients_tw（layer=0） 但是运行save_gradients_tw（layer=0）就需要运行weights_dict = torch.load（os.path.join（'./saved_model'， 'wingsnet_4.pth'）） 可是当前我们是训练过程，是没法得到这个save_model的，不是训练结束后才能得到model么？ 您看可不可以帮忙解答一下~非常感谢您~


冒昧打扰了.我在第一阶段的训练中，train.py 文件有一行 File “train.py”， line 244， in test weights_dict = torch.load（os.path.join（'./saved_model'， folder_name， 'wingsnet_18.pth'）），报错] 没有这样的文件或目录： './saved_model/0816_adddice/wingsnet_18.pth'，请问这个wingsnet_18.pth是如何获得的？您是否遇到过这样的问题？期盼您的回复，感激不尽
